### PR TITLE
implement winning conditions, find bug in capture logic

### PR DESCRIPTION
**Pull request by grezisek1**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
Based on the task requirements and the provided code changes, here is the code review:

1. **Task Requirement 1: Define the winning conditions for the Checkers game. Player wins when the opponent has no remaining pieces on the board.**

The code changes do not seem to cover this requirement. The winning condition is defined in the `isGameOver` function in `logic.js` as when a player's score reaches 15. However, this does not necessarily mean that the opponent has no remaining pieces on the board. The score could reach 15 due to other game rules (e.g., bonus points). 

A more accurate way to check if a player has won would be to iterate over the game board and check if there are any pieces belonging to the opponent. If there are none, then the current player has won.

```javascript
isGameOver(game) {
    for (let i = 0; i < game.state.data.length; i++) {
        if (game.state.data[i] == opponentPiece) {
            return false;
        }
    }
    return true;
}
```

2. **Task Requirement 2: Implement the logic to check for these winning conditions during the game.**

The code changes partially cover this requirement. The `isGameOver` function is called in `controller.js` after a piece is taken. However, it should also be called after a piece is moved without taking any opponent's piece. This is because a player could win not only by taking an opponent's piece but also by trapping the opponent so they cannot make a valid move.

```javascript
if (taken) {
    game.state.data[taken] = stateValues.empty;
    game.state.score[game.state.currentPlayerIndex]++;
}
if (game.logic.isGameOver(game)) {
    this.#endGame(game);
}
this.#movePiece(game, di);
```

3. **Task Requirement 3: Display a victory message when a player wins, indicating which player has won (Player 1 or Player 2).**

The code changes do not cover this requirement. There is no code to display a victory message. The `#endGame` function is called when the game is over, but it is not shown in the provided code changes. If it does not already do so, it should be updated to display a victory message indicating which player has won.

```javascript
#endGame(game) {
    let winningPlayer = game.state.currentPlayerIndex == 0 ? "Player 1" : "Player 2";
    alert(`${winningPlayer} has won the game!`);
}
```

In conclusion, the code changes do not fully cover the task requirements. The winning conditions are not correctly defined, the check for game over is not always performed when it should be, and there is no code to display a victory message.

**Comment by grezisek1** [comment]
> The score could reach 15 due to other game rules (e.g., bonus points).

It cant. There are no such bonuses.

> This is because a player could win not only by taking an opponent's piece but also by trapping the opponent so they cannot make a valid move.

That is true, but "Player wins when the opponent has no remaining pieces on the board." and trapped pieces are still on the board. Other game rules are also simplified.

> There is no code to display a victory message.

Look at `setWinner()` and `showWinner()` in the `UI` class, used from the `#endGame()` method.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
Based on the provided task requirements and the code changes, here is the review:

1. **Define the winning conditions for the Checkers game. Player wins when the opponent has no remaining pieces on the board.**

In the provided code changes, the winning condition is defined in the `isGameOver` function in `logic.js` file. However, the condition checks if the score of the current player is equal to 15. This does not align with the requirement that states a player wins when the opponent has no remaining pieces on the board. 

```diff
+    isGameOver(game) {
+        return game.state.score[game.state.currentPlayerIndex] == 15;
+    }
```

The winning condition should be defined as a check if the opponent has no remaining pieces on the board. This could be implemented by checking if the total number of opponent's pieces in `game.state.data` is zero.

2. **Implement the logic to check for these winning conditions during the game.**

The logic to check for winning conditions during the game is implemented in `controller.js` file. After a piece is taken, it checks if the game is over by calling `game.logic.isGameOver(game)`. This aligns with the requirement.

```diff
+                if (game.logic.isGameOver(game)) {
+                    this.#endGame(game);
+                }
```

3. **Display a victory message when a player wins, indicating which player has won (Player 1 or Player 2).**

The provided code changes do not include any code for displaying a victory message when a player wins. This requirement is not covered in the provided code changes.

In conclusion, the provided code changes partially cover the task requirements. The winning condition needs to be defined correctly according to the requirements and code for displaying a victory message needs to be added.

**Comment by grezisek1** [comment]
> This does not align with the requirement that states a player wins when the opponent has no remaining pieces on the board.

The score is a number of taken pieces from 15 available, so this is interchangeable. Checking the score avoids looping.

**Review by Sumick — APPROVED** [review]
